circleguard/visualizer.py

```diff
@@ -80,219 +129,242 @@ def next_frame_from_timer(self):
-        for i in range(len(self.buffer[index])-1):
-            self.draw_line(painter, self.CURSOR_COLORS[index], i*alpha_step, (self.buffer[index][i][1], self.buffer[index][i][2]), (self.buffer[index][i+1][1], self.buffer[index][i+1][2]))
-            self.draw_point(painter, self.CURSOR_COLORS[index], i*alpha_step, (self.buffer[index][i][1], self.buffer[index][i][2]))
-            if i == len(self.buffer[index])-2:
-                self.draw_point(painter, self.CURSOR_COLORS[index], (i+1)*alpha_step, (self.buffer[index][i+1][1], self.buffer[index][i+1][2]))
-
-    def paint_beatmap(self, painter):
+        alpha_step = 1 / FRAMES_ON_SCREEN
+        _pen = player["cursor_color"]
+        _pen.setWidth(WIDTH_LINE)
+        self.painter.setPen(_pen)
+        for i in range(len(player["buffer"]) - 1):
+            self.draw_line(i * alpha_step, (player["buffer"][i][1], player["buffer"][i][2]),
+                                           (player["buffer"][i + 1][1], player["buffer"][i + 1][2]))
+        _pen.setWidth(WIDTH_POINT)
+        self.painter.setPen(_pen)
+        for i in range(len(player["buffer"]) - 1):
+            self.draw_point(i * alpha_step, (player["buffer"][i][1], player["buffer"][i][2]))
+            if i == len(player["buffer"]) - 2:
+                self.draw_point((i + 1) * alpha_step, (player["buffer"][i + 1][1], player["buffer"][i + 1][2]))
+        # reset alpha
+        self.painter.setOpacity(1)
+
+    def paint_beatmap(self):
         for hitobj in self.hitobjs[::-1]:
-            self.draw_hitobject(painter, hitobj)
+            self.draw_hitobject(hitobj)
 
-    def paint_info(self, painter):
+    def paint_info(self):
         """
         Draws various Information.
 
         Args:
            QPainter painter: The painter.
         """
-        _pen = painter.pen()
-        painter.drawText(0, 15, f"Clock: {round(self.clock.get_time())} ms")
+        _pen = self.painter.pen()
+        self.painter.setOpacity(0.25)
+        self.painter.drawRect(X_OFFSET, Y_OFFSET, 512, 384)
+        self.painter.setOpacity(1)
+        self.painter.drawText(5, 15, f"Clock: {round(self.clock.get_time())} ms | Cursor count: {len(self.players)}")
         if self.replay_amount > 0:
-            for i in range(self.replay_amount):
-                painter.setPen(self.CURSOR_COLORS[i])
-                if len(self.buffer[i]) > 0:  # skips empty buffers
-                    painter.drawText(0, 30+(15*i), f"Cursor {self.usernames[i]}: {int(self.buffer[i][-1][1])}, {int(self.buffer[i][-1][2])}")
+            for i in range(len(self.players)):
+                player = self.players[i]
+                p = player["cursor_color"]
+                self.painter.setPen(QColor(0, 0, 0, 0))
+                self.painter.setBrush(QBrush(p.color()))
+                if len(self.players[i]["buffer"]) > 0:  # skips empty buffers
+                    self.painter.setOpacity(1 if Keys.M1 in Keys(int(player["buffer"][-1][3])) else 0.3)
```

Comment:
why only check `Keys.M1` and not `Keys.K1` as well?

Reply:
K1 is always a combination of K1 and M1, hence why you only need to check for M1.